File: Density/app/src/main/java/org/cornelldti/density/density/Facility.java
package org.cornelldti.density.density;

import java.io.Serializable;

public class Facility implements Serializable, Comparable<Facility> {
    private String name;
    private String id;
    public int occupancy_rating;
    private String description;
    private long closingAt;

    public void setClosingAt(long closingAt) {
        this.closingAt = closingAt;
    }

    public int getDensityResId() {
        if (!this.isOpen()) {
            return R.string.closed;
        }

        switch (this.occupancy_rating) {
            case 0:
                return R.string.very_empty;
            case 1:
                return R.string.pretty_empty;
            case 2:
                return R.string.pretty_crowded;
            case 3:
                return R.string.very_crowded;
        }

        return R.string.unknown;
    }

    @Override
    public int compareTo(Facility o) {
        if (!o.isOpen() && this.isOpen()) {
            return -1;
        }

        if (o.isOpen() && !this.isOpen()) {
            return 1;
        }

        if (o.getOccupancyRating() < this.getOccupancyRating()) {
            return 1;
        }

        if (o.getOccupancyRating() > this.getOccupancyRating()) {
            return -1;
        }

        return 0;
    }

    public enum CampusLocation {NORTH, WEST, CENTRAL}

    private CampusLocation loc;

    public Facility(String name, String id) {
        this.name = name;
        this.id = id;
    }

    public Facility(String name, String id, String description, long nextOpen, long closingAt, String address, CampusLocation campusLocation, int occupancy_rating) {
        this.name = name;
        this.id = id;
        this.description = description;
        this.loc = campusLocation;
        this.occupancy_rating = occupancy_rating;
        this.closingAt = closingAt;
    }

    public boolean isOpen() {
        return this.closingAt != -1;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getName() {
        return this.name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getId() {
        return this.id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getDescription() {
        return this.description;
    }

    public int getOccupancyRating() {
        return occupancy_rating;
    }

    public Facility setOccupancy_rating(int i) {
        if (i >= 0 && i <= 3) {
            this.occupancy_rating = i;
        }
        return this;
    }

    public CampusLocation getLocation() {
        return this.loc;
    }

    public String getLocationString() {
        String loc_string = "";
        switch (this.loc) {
            case NORTH:
                loc_string = "NORTH";
                break;
            case CENTRAL:
                loc_string = "CENTRAL";
                break;
            case WEST:
                loc_string = "WEST";
                break;
        }
        return loc_string;
    }

    public Facility setLocation(String loc) {
        switch (loc) {
            case "north":
                this.loc = CampusLocation.NORTH;
                break;
            case "central":
                this.loc = CampusLocation.CENTRAL;
                break;
            case "west":
                this.loc = CampusLocation.WEST;
                break;
        }
        return this;
    }
}

File: Density/app/src/main/java/org/cornelldti/density/density/Facility.kt
package org.cornelldti.density.density

import java.io.Serializable

class Facility : Serializable, Comparable<Facility> {
    var name: String? = null
    var id: String? = null
    var occupancyRating: Int = 0
    var description: String? = null
    private var closingAt: Long = 0

    val densityResId: Int
        get() {
            if (!this.isOpen) {
                return R.string.closed
            }

            when (this.occupancyRating) {
                0 -> return R.string.very_empty
                1 -> return R.string.pretty_empty
                2 -> return R.string.pretty_crowded
                3 -> return R.string.very_crowded
            }

            return R.string.unknown
        }

    var location: CampusLocation? = null
        private set

    val isOpen: Boolean
        get() = this.closingAt != -1L

    val locationString: String
        get() = when (this.location) {
            CampusLocation.NORTH -> "NORTH"
            CampusLocation.CENTRAL -> "CENTRAL"
            CampusLocation.WEST -> "WEST"
            else -> ""
        }

    fun setClosingAt(closingAt: Long) {
        this.closingAt = closingAt
    }

    override fun compareTo(o: Facility): Int {
        if (!o.isOpen && this.isOpen) {
            return -1
        }

        if (o.isOpen && !this.isOpen) {
            return 1
        }

        if (o.occupancyRating < this.occupancyRating) {
            return 1
        }

        return if (o.occupancyRating > this.occupancyRating) {
            -1
        } else 0

    }

    enum class CampusLocation {
        NORTH, WEST, CENTRAL
    }

    constructor(name: String, id: String) {
        this.name = name
        this.id = id
    }

    constructor(name: String, id: String, description: String, nextOpen: Long, closingAt: Long, address: String, campusLocation: CampusLocation, occupancy_rating: Int) {
        this.name = name
        this.id = id
        this.description = description
        this.location = campusLocation
        this.occupancyRating = occupancy_rating
        this.closingAt = closingAt
    }

    fun setOccupancy_rating(i: Int): Facility {
        if (i in 0..3) {
            this.occupancyRating = i
        }
        return this
    }

    fun setLocation(loc: String): Facility {
        when (loc) {
            "north" -> this.location = CampusLocation.NORTH
            "central" -> this.location = CampusLocation.CENTRAL
            "west" -> this.location = CampusLocation.WEST
        }
        return this
    }
}
